server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,479 @@
+                    state.segmentSuffix
+                );
+                final PostingsFormat postingsFormat = forName(metaIn.readString());
+                final int numBloomFilters = metaIn.readVInt();
+                bloomFilters = new HashMap<>(numBloomFilters);
+                for (int i = 0; i < numBloomFilters; i++) {
+                    final int fieldNumber = metaIn.readVInt();
+                    final long startFilePointer = metaIn.readVLong();
+                    final int bloomFilterSize = metaIn.readVInt();
+                    assert bloomFilterSize == bloomFilterSize(state.segmentInfo.maxDoc())
+                        : "bloom_filter_size=" + bloomFilterSize + ", max_docs=" + state.segmentInfo.maxDoc();
+                    final String fieldName = state.fieldInfos.fieldInfo(fieldNumber).name;
+                    bloomFilters.put(fieldName, new BloomFilter(fieldName, startFilePointer, bloomFilterSize));
+                }
+                CodecUtil.checkFooter(metaIn);
+                termsReader = postingsFormat.fieldsProducer(state);
+                toCloses.add(this.termsReader);
+                indexIn = state.directory.openInput(indexFile(state.segmentInfo, state.segmentSuffix), state.context);
+                toCloses.add(indexIn);
+                CodecUtil.checkIndexHeader(
+                    indexIn,
+                    BLOOM_CODEC_NAME,
+                    VERSION_START,
+                    VERSION_CURRENT,
+                    state.segmentInfo.getId(),
+                    state.segmentSuffix
+                );
+                CodecUtil.retrieveChecksum(indexIn);
+                toCloses.clear();
+            } finally {
+                IOUtils.closeWhileHandlingException(toCloses);
+            }
+        }
+
+        @Override
+        public Iterator<String> iterator() {
+            return termsReader.iterator();
+        }
+
+        @Override
+        public void close() throws IOException {
+            IOUtils.close(termsReader, indexIn);
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            final Terms terms = termsReader.terms(field);
+            if (terms == null) {
+                return null;
+            }
+            final BloomFilter bloomFilter = bloomFilters.get(field);
+            if (bloomFilter != null) {
+                final RandomAccessInput data = indexIn.randomAccessSlice(
+                    bloomFilter.startFilePointer(),
+                    numBytesForBloomFilter(bloomFilter.bloomFilterSize)
+                );
+                return new BloomFilterTerms(terms, data, bloomFilter.bloomFilterSize);
+            } else {
+                return terms;
+            }
+        }
+
+        @Override
+        public int size() {
+            return termsReader.size();
+        }
+
+        @Override
+        public void checkIntegrity() throws IOException {
+            termsReader.checkIntegrity();
```

Comment:
Let's also do a `checksumEntireFile` on the file that holds the bloom filter? (No need to check the meta file since it's checked upon opening)